# Indel_genotyping
A tool to genotype indels generated by multiplexed CRISPR-Cas9 using amplicon-sequencing. The script scans for indels up to 20bp and reports indels found in at least 10% of all sequenced reads (frequency > 10%). Substitutions are ignored.

## Usage
./amplicon_seq_pipeline.sh -q fastq_directory -f fasta_file.fa -n study_name  
### Options
-q directory to your raw sequencing data in fq.gz format. If your file has a lot of useless stuff in its name, it will be kept in this pipeline. For clarity, use something like sampleID.fq.gz  
-f fasta file that contains your target genes (relative or absolute path).  
-n study name, a prefix that will be added to files and folders.  
The command should be run from the directory where the script files are located. The program is meant to be run in interactive mode on the cluster but should work fine as a slurm job if you add a slurm header to the main script.  
Maximum tested indel size and threshold for reported indels can be customized easily by changing argument values at the beginning of *amplicon_seq_pipeline.sh*  

## Output files
All output files are in a directory with your study name, located in the current directory, that contains the following:  
* **total_frequency_of_indels.tsv** : for each sample and gene, adds up the frequencies of all indels that passed the frequency threshold (indel found in > 10% of the reads by default). Table is sorted from sample with highest sum of all frequencies to lowest. This table is used to select samples with the lowest frequency of WT alleles, but does not take into account whether indels are fixed (single indel with high frequency, typically close to 50% or 100%) or if they are a mixture of many different low-frequency mutations. I would recommend using this table to select lines rather than the file below.
* **frequency_of_top_coverage_indels.tsv** : for each sample and gene, selects the indel with the highest **coverage** (we would rather like highest frequency here, but highest frequency mutations often are low coverage mutations that we are not interested in) but reports its **frequency**. Table is sorted from sample with highest sum of all frequencies to lowest.
* **idxstats.tsv** Aggregated output from samtools idxstats. Reports the gene size in 1st column followed by number of reads at each gene for each sample. Used for quality control, coverage should be relatively homogeneous.
* **bam**: folder with the aligned reads (.bam) and index files (.bai). Needed to vizualize reads in IGV or another genome browser
* **mutations**: folder with all indels above the frequency threshold (10% by default) for each gene. Columns are: genename, position, reference sequence, coverage, indel, frequency
* bwa_logs, indexes, mpileup folders : intermediate folders you can ignore

## Analysis
Workflow is as follow: use .tsv tables to find samples with the most indels. Check the most interesting samples in IGV by dragging bam files. The reason to check is that the script only tests indels up to 20bp, so substitutions and larger indels are missed (not so rare). IGV also helps to vizualize alleles with complex mutations.

